### ISSUE TITLE
Namespaces instrumented in incorrect order

### ISSUE THREAD

**Issue by camsaul** [cloverage/cloverage#302]
I ran into this while debugging some false negatives in https://github.com/metabase/metabase.

Suppose we have three namespaces: 

<details>
<summary>
<tt>a</tt>
</summary>

```clj
(ns cloverage-test.a
  (:require [cloverage-test.handler :as handler]))

(defn process-request [m]
  (handler/handler (assoc m :request? true)))
```
</details>

<details>
<summary>
<tt>handler</tt>
</summary>

```clj
(ns cloverage-test.handler
  (:require [cloverage-test.handler.middleware :as middleware]))

(def middleware
  [middleware/my-middleware])

(def handler
  (reduce
   (fn [handler middleware]
     (middleware handler))
   identity
   middleware))
```

</details>

<details>
<summary>
<tt>middleware</tt>
</summary>

```clj
(ns cloverage-test.handler.middleware)

(defn my-middleware [handler]
  (fn [query]
    (handler (assoc query :middleware? true))))
```

</details>

Loading `a` will load `handler` which in turn will load `middleware`. 

<details>
<summary>
Suppose we also have a single test that covers everything.
</summary>

```clj
(ns cloverage-test.core-test
  (:require [clojure.test :refer :all]
            [cloverage-test.a :as a]))

(deftest process-request-test
  (is (= {:request? true, :middleware? true}
         (a/process-request {}))))
```

</details>

### Here's where things get wacky.

Now suppose we want to instrument `a` and `middleware`. 

<details>
<summary>
If Cloverage attempts to instrument <tt>a</tt> first, <tt>middleware</tt> and <tt>handler</tt> get loaded as a side effect, and forms therein are evaluated; because they are not yet instrumented, these forms are not recorded as having been evaluated. If these namespaces are instrumented later they are falsely recorded as never having been evaluated at all
</summary>

```
Loading namespaces:  (cloverage-test.a cloverage-test.handler.middleware)
Test namespaces:  (cloverage-test.core-test)
[Loading cloverage-test.handler.middleware]
cloverage.coverage/*covered*: bound? true identity hash code: 1890212065
Stacktrace:
("cloverage.instrument$eval_form$fn__3001.invoke(instrument.clj:474)"
 "cloverage.instrument$eval_form.invokeStatic(instrument.clj:472)"
 "cloverage.instrument$eval_form.invoke(instrument.clj:468)"
 "cloverage.instrument$instrument_form.invokeStatic(instrument.clj:510)"
 "cloverage.instrument$instrument_form.invoke(instrument.clj:496)"
 "cloverage.instrument$instrument_file.invokeStatic(instrument.clj:527)"
 "cloverage.instrument$instrument_file.invoke(instrument.clj:523)"
 "cloverage.instrument$instrument.invokeStatic(instrument.clj:538)"
 "cloverage.instrument$instrument.invoke(instrument.clj:533)"
 "cloverage.coverage$instrument_namespaces$fn__4093.invoke(coverage.clj:160)"
 "cloverage.coverage$instrument_namespaces.invokeStatic(coverage.clj:155)"
 "cloverage.coverage$instrument_namespaces.invoke(coverage.clj:151)"
 "cloverage.coverage$run_main.invokeStatic(coverage.clj:260)"
 "cloverage.coverage$run_main.invoke(coverage.clj:253)"
 "cloverage.coverage$run_project.invokeStatic(coverage.clj:277)"
 "cloverage.coverage$run_project.doInvoke(coverage.clj:273)")
Loaded  cloverage-test.a  .
[Loading cloverage-test.handler.middleware]
cloverage.coverage/*covered*: bound? true identity hash code: 342419584
Stacktrace:
("cloverage.instrument$eval_form$fn__3001.invoke(instrument.clj:474)"
 "cloverage.instrument$eval_form.invokeStatic(instrument.clj:472)"
 "cloverage.instrument$eval_form.invoke(instrument.clj:468)"
 "cloverage.instrument$instrument_form.invokeStatic(instrument.clj:510)"
 "cloverage.instrument$instrument_form.invoke(instrument.clj:496)"
 "cloverage.instrument$instrument_file.invokeStatic(instrument.clj:527)"
 "cloverage.instrument$instrument_file.invoke(instrument.clj:523)"
 "cloverage.instrument$instrument.invokeStatic(instrument.clj:538)"
 "cloverage.instrument$instrument.invoke(instrument.clj:533)"
 "cloverage.coverage$instrument_namespaces$fn__4093.invoke(coverage.clj:160)"
 "cloverage.coverage$instrument_namespaces.invokeStatic(coverage.clj:155)"
 "cloverage.coverage$instrument_namespaces.invoke(coverage.clj:151)"
 "cloverage.coverage$run_main.invokeStatic(coverage.clj:260)"
 "cloverage.coverage$run_main.invoke(coverage.clj:253)"
 "cloverage.coverage$run_project.invokeStatic(coverage.clj:277)"
 "cloverage.coverage$run_project.doInvoke(coverage.clj:273)")
Loaded  cloverage-test.handler.middleware  .
Instrumented namespaces.

Testing cloverage-test.core-test

Ran 1 tests containing 1 assertions.
0 failures, 0 errors.
Ran tests.
Writing HTML report to: /home/cam/cloverage-test/target/coverage/index.html

|-----------------------------------+---------+---------|
|                         Namespace | % Forms | % Lines |
|-----------------------------------+---------+---------|
|                  cloverage-test.a |  100.00 |  100.00 |
| cloverage-test.handler.middleware |   83.08 |   83.33 |
|-----------------------------------+---------+---------|
|                         ALL FILES |   85.14 |   86.67 |
|-----------------------------------+---------+---------|
```

![image](https://user-images.githubusercontent.com/1455846/90281347-41f3b880-de21-11ea-93f1-16e4e61dd305.png)
</details>

<details>
<summary>
If Cloverage instruments <tt>middleware</tt> first coverage is reported correctly.
</summary>

```
Loading namespaces:  (cloverage-test.a cloverage-test.handler cloverage-test.handler.middleware)
Test namespaces:  (cloverage-test.core-test)
[Loading cloverage-test.handler.middleware]
cloverage.coverage/*covered*: bound? true identity hash code: 686366573
Stacktrace:
("cloverage.instrument$eval_form$fn__3001.invoke(instrument.clj:474)"
 "cloverage.instrument$eval_form.invokeStatic(instrument.clj:472)"
 "cloverage.instrument$eval_form.invoke(instrument.clj:468)"
 "cloverage.instrument$instrument_form.invokeStatic(instrument.clj:510)"
 "cloverage.instrument$instrument_form.invoke(instrument.clj:496)"
 "cloverage.instrument$instrument_file.invokeStatic(instrument.clj:527)"
 "cloverage.instrument$instrument_file.invoke(instrument.clj:523)"
 "cloverage.instrument$instrument.invokeStatic(instrument.clj:538)"
 "cloverage.instrument$instrument.invoke(instrument.clj:533)"
 "cloverage.coverage$instrument_namespaces$fn__4093.invoke(coverage.clj:160)"
 "cloverage.coverage$instrument_namespaces.invokeStatic(coverage.clj:155)"
 "cloverage.coverage$instrument_namespaces.invoke(coverage.clj:151)"
 "cloverage.coverage$run_main.invokeStatic(coverage.clj:260)"
 "cloverage.coverage$run_main.invoke(coverage.clj:253)"
 "cloverage.coverage$run_project.invokeStatic(coverage.clj:277)"
 "cloverage.coverage$run_project.doInvoke(coverage.clj:273)")
Loaded  cloverage-test.handler.middleware  .
Loaded  cloverage-test.handler  .
Loaded  cloverage-test.a  .
Instrumented namespaces.

Testing cloverage-test.core-test

Ran 1 tests containing 1 assertions.
0 failures, 0 errors.
Ran tests.
Writing HTML report to: /home/cam/cloverage-test/target/coverage/index.html

|-----------------------------------+---------+---------|
|                         Namespace | % Forms | % Lines |
|-----------------------------------+---------+---------|
|                  cloverage-test.a |  100.00 |  100.00 |
|            cloverage-test.handler |  100.00 |  100.00 |
| cloverage-test.handler.middleware |   95.38 |  100.00 |
|-----------------------------------+---------+---------|
|                         ALL FILES |   96.55 |  100.00 |
|-----------------------------------+---------+---------|
```

![image](https://user-images.githubusercontent.com/1455846/90281900-4a98be80-de22-11ea-810e-ff2e1eb174a5.png)

</details>

#### Project

I've pushed a GitHub project to demonstrate this bug: https://github.com/camsaul/cloverage-duplicate-instrumentation-project

You can trigger the bug by running Cloverage like

```bash
lein cloverage --ns-regex '^cloverage-test\.(?:(?:handler\.middleware)|a)$
```

Running without `--ns-regex` will not trigger the bug:

```bash
lein cloverage
```

It does seem that `--ns-regex` causes the namespaces to be loaded in a different order, but I don't think that's the root cause of this bug, as we were running into it with Metabase without using `--ns-regex`. `--ns-regex` is just the easiest way I've found to trigger this bug.

**Comment by camsaul** [comment]
I haven't dug into this bug yet but am planning on doing so now that I have a minimal project for reproducing it.

**Comment by camsaul** [comment]
So it seems like the correct way to fix this would be to always instrument namespaces before they get loaded as a side-effect, e.g. construct a dependency graph by reading the `:require` form in the `ns` declaration and then instrument those namespaces first (if needed)

**Comment by camsaul** [comment]
It seems like this is the way things are supposed to work:

```clj
(defn load-namespaces [{:keys [extra-test-ns ns-regex test-ns-regex ns-exclude-regex src-ns-path test-ns-path]} add-nses]
  (let [include      (-> src-ns-path
                         (find-nses ns-regex)
                         (remove-nses ns-exclude-regex))
        namespaces   (concat add-nses include)
        test-nses    (concat extra-test-ns (find-nses test-ns-path test-ns-regex))
        ordered-nses (dep/in-dependency-order (map symbol namespaces))]
    ...
    {:test-nses test-nses, :ordered-nses ordered-nses}))
```

There must be a bug with `dep/in-dependency-order`

**Comment by camsaul** [comment]
So I actually discovered a couple bugs. 

1.  `load-namespaces` only determines dependency order for namespaces that are getting instrumented, and doesn't take into account dependency order for other namespaces. Thus if you only want to instrument a handful of namespaces the order likely won't be correct. The fix for this is to take every namespace into account when determine order.

2.  The dependency ordering code is not working correctly even when attempting to instrument every namespace in the project. I'm the maintainer of https://github.com/metabase/metabase which is a pretty big project (about 320 namespaces) and I can say with 100% certainty that the order `cloverage.dependency/in-dependency-order` came up with is not correct.

PR forthcoming.

**Comment by camsaul** [comment]
Fixed by #303
